sw/device/lib/runtime/pmp.h

```diff
@@ -0,0 +1,312 @@
+ * "3.6.1 Physical Memory Protection CSRs", "Address Matching".
+ *
+ * PMP regions matching logic and prioritisation is described in the
+ * "Priority and Matching Logic" paragraph of the section "3.6.1".
+ *
+ * Note that granularity is implementation defined (default G=0, making the
+ * minimal PMP region size as 2^(G+2) = 4 bytes), and with G ≥ 1, the
+ * NA4 is not available. Granularity also determines the minimal region size in
+ * TOR and NAPOT modes.
+ */
+
+/**
+ * Attribute for functions which return errors that must be acknowledged.
+ *
+ * This attribute must be used to mark all PMP API that return an error value of
+ * some kind, to ensure that callers do not accidentally drop the error on the
+ * ground.
+ */
+#define PMP_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
+
+/**
+ * RV32 PMP CSR definitions.
+ *
+ * The values for these defines have been taken from the:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6.1 Physical Memory Protection CSRs".
+ */
+#define PMP_REGIONS_NUM 16
+#define PMP_CFG_MULTIREG_COUNT 4
+#define PMP_ADDR_REG_COUNT 16
+
+/**
+ * Ibex PMP address granularity.
+ *
+ * Please see:
+ * https://ibex-core.readthedocs.io/en/latest/pmp.html#pmp-granularity .
+ */
+#define PMP_GRANULARITY_IBEX 0
+
+/**
+ * RISC-V PMP address alignment and granularity.
+ *
+ * The values have been taken from the:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6 Physical Memory Protection", section "Address Matching".
+ *
+ * "PMP mechanism supports regions as small as four bytes, platforms may
+ * specify coarser PMP regions."
+ */
+#define PMP_ADDRESS_MIN_ALIGNMENT 4
+#define PMP_ADDRESS_MIN_ALIGNMENT_NAPOT 8
+#define PMP_ADDRESS_ALIGNMENT (4 << PMP_GRANULARITY_IBEX)
+#define PMP_ADDRESS_ALIGNMENT_MASK (PMP_ADDRESS_ALIGNMENT - 1)
+#define PMP_ADDRESS_ALIGNMENT_INVERTED_MASK (~PMP_ADDRESS_ALIGNMENT_MASK)
+#define PMP_ADDRESS_SHIFT 2
+
+/**
+ * PMP region access permissions.
+ *
+ * PMP permissions are described in:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6.1 Physical Memory Protection CSRs", "Locking and Privilege Mode".
+ *
+ * Note that "The combination R=0 and W=1 is reserved for future use".
+ *
+ * When a region is configured with `kPmpRegionLockUnlocked` then these
+ * permissions only apply to RISC-V "U" and "S" privilege modes, and have no
+ * effect in the "M" mode.
+ */
+typedef enum pmp_region_permissions {
+  kPmpRegionPermissionsNone = 0,         /**< No access permitted .*/
+  kPmpRegionPermissionsReadOnly,         /**< Only read access. */
+  kPmpRegionPermissionsExecuteOnly,      /**< Only execute access. */
+  kPmpRegionPermissionsReadExecute,      /**< Read and execute access. */
+  kPmpRegionPermissionsReadWrite,        /**< Read and write access. */
+  kPmpRegionPermissionsReadWriteExecute, /**< Read, write and execute access. */
```

Comment:
Yes, that is correct. Miguel has also pointed it out that this could be done with bitfields. I can always check for an illegal combination in the source instead.

I think I will leave this refactor for the follow-up.